From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Sleigh-InSPECtor <sleighinspector@outlook.com>
Date: Thu, 16 May 2024 12:08:49 +0930
Subject: [PATCH] 6529: AArch32: sha1su0.32 had a destructive 64-bit left shift
 on a 64-bit value before zext

AArch32: fixed sha1su0.32

* fixed destructive 64-bit left shift on a 64-bit value before zext
---
 Ghidra/Processors/ARM/data/languages/ARMneon.sinc | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/Ghidra/Processors/ARM/data/languages/ARMneon.sinc b/Ghidra/Processors/ARM/data/languages/ARMneon.sinc
index eb0da7014a..f558f7f921 100644
--- a/Ghidra/Processors/ARM/data/languages/ARMneon.sinc
+++ b/Ghidra/Processors/ARM/data/languages/ARMneon.sinc
@@ -619,9 +619,9 @@ define pcodeop SHA1HashUpdateParity;
 	local op1 = Qd;
 	local op2 = Qn;
 	local op3 = Qm;
-	local op2lo:8 = op2(0);
-	local op1hi:8 = op1(8);
-	op2 = zext(op2lo << 64) | zext(op1hi);
+	local op2LowerHalf = zext(op2[0,64]) << 64;
+	local op1UpperHalf = zext(op1[64,64]);
+	op2 = op2LowerHalf | op1UpperHalf;
 	Qd = op1 ^ op2 ^ op3;
 }
 
-- 
2.45.1

